viewsb/packet.py

```diff
@@ -733,9 +758,133 @@ def from_setup_data(cls, setup_data, **fields):
+        elif descType == 7:
+            return 'other speed'
+        elif descType == 8:
+            return 'interface power'
+        elif descType == 9:
+            return 'otg'
+        elif descType == 10:
+            return 'debug'
+        elif descType == 11:
+            return 'interface association'
+        elif descType == 12:
+            return 'security'
+        elif descType == 13:
+            return 'key'
+        elif descType == 14:
+            return 'encryption type'
+
+        elif descType == 16:
+            return 'device capability'
+        elif descType == 17:
+            return 'wireless endpoint'
+
+        elif descType == 33:
+            return 'hid'
+        elif descType == 34:
+            return 'hid report'
+        elif descType == 35:
+            return 'physical description'
+        return f'unknown ({descType:#x})'
 
-    def summarize_data(self, summary_length_bytes=8):
-        return "value={:04x} index={:04x} length={:04x}".format(self.value, self.index, self.request_length)
+    @staticmethod
+    def _decode_descriptor_lang(lang):
+        if lang == 0x0904:
+            return 'english'
+        return f'unknown {lang:#06x}'
+
+    @classmethod
+    def _decode_descriptor_request(cls, data):
+        index = data.wValue & 0xFF
+        descType = data.wValue >> 8
+        data.wValue = {
+            'Index': index,
+            'Type': cls._decode_descriptor_type(descType),
+        }
+        if data.wIndex != 0:
+            data.wIndex = {
+                'Language': cls._decode_descriptor_lang(data.wIndex),
+            }
+
+    @staticmethod
+    def _decode_request_type(reqType):
+        if reqType == 0:
+            return 'standard'
+        elif reqType == 1:
+            return 'class'
+        elif reqType == 2:
+            return 'vendor'
+        return 'reserved'
+
+    @staticmethod
+    def _decode_request_recipient(recipient):
+        if recipient == 0:
+            return 'device'
+        elif recipient == 1:
+            return 'interface'
+        elif recipient == 2:
+            return 'endpoint'
+        elif recipient == 3:
+            return 'other'
+        return 'reserved'
+
+    @classmethod
+    def _decode_request(cls, data):
+        if data.bRequest == 6:
+            cls._decode_descriptor_request(data)
+
+        reqDir = (data.bmRequestType >> 7)
+        reqType = (data.bmRequestType >> 5) & 0x3
+        reqRecipient = data.bmRequestType & 0x1F
+        data.bmRequestType = {
+            'Direction': 'device-to-host' if reqDir == 1 else 'host-to-device',
+            'Type': cls._decode_request_type(reqType),
+            'Recipient': cls._decode_request_recipient(reqRecipient),
+        }
+
+        data.bRequest = cls.REQUESTS.get(data.bRequest, f'UNKNOWN ({data.bRequest:#x})')
+
+    def get_detail_fields(self):
+        # If we don't have a parsed version of this class, try to parse it.
+        if not hasattr(self, 'parsed'):
```

Comment:
Generally, I think we should avoid this kind of introspection. I think a better solution would be setting initializing `self.parsed` to `None`.

Reply:
Ok, fair comment.. this needs to be carefully re-thought then as this is done to prevent over-parsing packets being parsed in other places in the code base (standard descriptors decoders).

Reply:
Just checked, and yeah.. this was done to match the way it works in https://github.com/usb-tools/ViewSB/blob/master/viewsb/descriptor.py#L204 - I suggest both need revisiting/revising but that perhaps that should be a separate PR as that requires some careful rewriting of the decoding and caching process.